src/libsyntax/parse/parser.rs

```diff
@@ -3618,7 +3618,14 @@ impl<'a> Parser<'a> {
     {
         let (iattrs, body) = self.parse_inner_attrs_and_block()?;
         attrs.extend(iattrs);
-        Ok(self.mk_expr(span_lo.to(body.span), ExprKind::TryBlock(body), attrs))
+        if self.eat_keyword(keywords::Catch) {
+            let mut error = self.struct_span_err(self.prev_span,
+                                                 "`try {} catch` is not a valid syntax");
+            error.help("try using `match` on the result of the `try` block instead");
+            Err(error)
```

Comment:
I'd be tempted at consuming the entire `catch` block and returning `Ok`... But this is safer, at least for now.

Reply:
Yeah I prefer not trying recovery here for now in case it turns out buggy.